### PR TITLE
Deprecate document_id_field and add support for document_id with formatting

### LINKED ISSUE #3074: Support formatting of the document_id_field in the OpenSearch sink
**Is your feature request related to a problem? Please describe.**

In the `index` parameter of the OpenSearch sink this format is supported to format from multiple keys or expression results in the Event

```
index: "${key-one}-${key-two}-${getMetadata(\"some_metadata_key\")"
```

but `document_id_field` has never supported the format option, so it has to either be

```
document_id_field: "some_key"
```

or

```
document_id_field: "getMetadata(\"some_metadata_key\")"
```

**Describe the solution you'd like**
To allow the same formatting as `index` for `document_id_field`, but without breaking the existing one (if a temporary `document_id_field_format` option is needed until it is possible to make a breaking change, then that works as well. It may be possible to do now without breaking the existing functionality though 

```
document_id_field: "${key-one}-${key-two}-${getMetadata(\"some_metadata_key\")}"
```

**Describe alternatives you've considered (Optional)**

**Additional context**

### PR DESCRIPTION
**Pull request by graytaylor0**
### Description
Instead of supporting an expression or an Event key field in `document_id_field` parameter, we have deprecated `document_id_field` and added a new parameter `document_id`, which will support formatting similar to `index`. This format can take Event key names or expressions. 

Ex:

```
document_id: "test-${getMetadata(\"some_metadata_key\"}-${some_key}
```

Tested sending data with both `document_id_field` and `document_id`, and both work as intended
 
### Issues Resolved
Resolves #3074 

### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexConfiguration.java [hunk 1]
```diff
@@ -119,12 +127,16 @@ private IndexConfiguration(final Builder builder) {
         this.routingField = builder.routingField;
 
         String documentIdField = builder.documentIdField;
+        String documentId = builder.documentId;
         if (indexType.equals(IndexType.TRACE_ANALYTICS_RAW)) {
             documentIdField = "spanId";
+            documentId = "spanId";
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexConfiguration.java:133]: Should you remove the `documentIdField` from above?

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexConfiguration.java [hunk 2]
```diff
@@ -119,12 +127,16 @@ private IndexConfiguration(final Builder builder) {
...
         } else if (indexType.equals(IndexType.TRACE_ANALYTICS_SERVICE_MAP)) {
             documentIdField = "hashId";
+            documentId = "hashId";
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexConfiguration.java:136]: Again, shouldn't we remove `documentIdField = "hashId";`?

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexConfiguration.java (hunk 2 above)
Comment (graytaylor0) [reply data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexConfiguration.java:136]: That would be a breaking change wouldn't it?

Comment (graytaylor0) [reply data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexConfiguration.java:136]: Oh I see I'll add a check to keep the null one null

**Review by cmanning09 — COMMENTED** [review]
Let's we write unit tests for the IndexConfiguration class.

**Review by asifsmohammed — DISMISSED** [review]

**Comment by asifsmohammed** [comment]
What's the plan on updating docs in project website?

**Review by dlvenable — DISMISSED** [review]

**Review by asifsmohammed — APPROVED** [review]

**Review by dlvenable — APPROVED** [review]
Thanks for adding those tests!

**Pull request merged**